src/Microsoft.AspNetCore.SignalR.Core/Internal/HubMethodDescriptor.cs

```diff
@@ -0,0 +1,149 @@
+
+            if (IsObservableType(NonAsyncReturnType, out var observableItemType))
+            {
+                IsObservable = true;
+                StreamReturnType = observableItemType;
+            }
+            else if (IsChannelType(NonAsyncReturnType, out var channelItemType))
+            {
+                IsChannel = true;
+                StreamReturnType = channelItemType;
+            }
+        }
+
+        private Func<object, CancellationToken, IAsyncEnumerator<object>> _convertToEnumerator;
+
+        public ObjectMethodExecutor MethodExecutor { get; }
+
+        public IReadOnlyList<Type> ParameterTypes { get; }
+
+        public Type NonAsyncReturnType { get; }
+
+        public bool IsObservable { get; }
+
+        public bool IsChannel { get; }
+
+        public bool IsStreamable => IsObservable || IsChannel;
+
+        public Type StreamReturnType { get; }
+
+        public IList<IAuthorizeData> Policies { get; }
+
+        private static bool IsChannelType(Type type, out Type payloadType)
+        {
+            var channelType = type.AllBaseTypes().FirstOrDefault(t => t.IsGenericType && t.GetGenericTypeDefinition() == typeof(ChannelReader<>));
+            if (channelType == null)
+            {
+                payloadType = null;
+                return false;
+            }
+
+            payloadType = channelType.GetGenericArguments()[0];
+            return true;
+        }
+
+        private static bool IsObservableType(Type type, out Type payloadType)
+        {
+            var observableInterface = IsIObservable(type) ? type : type.GetInterfaces().FirstOrDefault(IsIObservable);
+            if (observableInterface == null)
+            {
+                payloadType = null;
+                return false;
+            }
+
+            payloadType = observableInterface.GetGenericArguments()[0];
+            return true;
+
+            bool IsIObservable(Type iface)
+            {
+                return iface.IsGenericType && iface.GetGenericTypeDefinition() == typeof(IObservable<>);
+            }
+        }
+
+        public IAsyncEnumerator<object> FromObservable(object observable, CancellationToken cancellationToken)
+        {
+            if (_convertToEnumerator == null)
+            {
+                _convertToEnumerator = CompileConvertToEnumerator(FromObservableMethod);
+            }
+
+            return _convertToEnumerator.Invoke(observable, cancellationToken);
+        }
+
+        public IAsyncEnumerator<object> FromChannel(object channel, CancellationToken cancellationToken)
+        {
+            if (_convertToEnumerator == null)
+            {
+                _convertToEnumerator = CompileConvertToEnumerator(GetAsyncEnumeratorMethod);
+            }
+
+            return _convertToEnumerator.Invoke(channel, cancellationToken);
+        }
+
+        private Func<object, CancellationToken, IAsyncEnumerator<object>> CompileConvertToEnumerator(MethodInfo methodInfo)
```

Comment:
Can this be static?

Reply:
`methodInfo` should perhaps be `adapterMethod` because I initially confused it with the hub method.